### PR TITLE
MGMT-7115: When running test_parallel, VIPs DNS cannot be resolved from host

### PR DESCRIPTION
**Pull request by ori-amizur**
- Created DNS forwarding file(s) in /etc/NetworkManager/dnsmasq.d by using terraform.

### CODE REVIEW

File: discovery-infra/test_infra/helper_classes/cluster.py [hunk 1]
```diff
@@ -621,6 +621,18 @@ def wait_for_install(self, timeout=consts.CLUSTER_INSTALLATION_TIMEOUT):
...
+        cluster = self.api_client.cluster_get(self.id)
+        cluster_name = cluster.name
+        base_domain = cluster.base_dns_domain
+        main_cidr = self.nodes.controller.get_machine_cidr()
+        nameserver_ip = str(IPNetwork(main_cidr).ip + 1)
+        fname = f"/etc/NetworkManager/dnsmasq.d/openshift-{cluster_name}.conf"
+        contents = f"server=/api.{cluster_name}.{base_domain}/{nameserver_ip}\n" + f"server=/.apps.{cluster_name}.{base_domain}/{nameserver_ip}\n"
+        self.nodes.controller.tf.change_variables(
+            {"dns_forwarding_file": contents, "dns_forwarding_file_name": fname}
+        )
```
Comment (osherdp) [inline discovery-infra/test_infra/helper_classes/cluster.py:634]: What's the reason for managing those files in terraform?

**Review by osherdp — CHANGES_REQUESTED** [review]

File: skipper.yaml [hunk 2]
```diff
@@ -35,6 +35,9 @@ volumes:
   - /tmp:/tmp/
   - /dev/:/dev
   - /run/udev:/run/udev
+
+  # Network manager dnsmasq
+  - /etc/NetworkManager/dnsmasq.d:/etc/NetworkManager/dnsmasq.d
```
Comment (YuviGold) [inline skipper.yaml:40]: Please add a comment on why we're mounting it

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 1 above)
Comment (YuviGold) [reply discovery-infra/test_infra/helper_classes/cluster.py:634]: +1
Isn't managing it with TF just make it more complex?

File: Makefile [hunk 3]
```diff
@@ -478,6 +478,7 @@ _test: $(REPORTS) _test_setup
 test_parallel:
 	$(MAKE) start_load_balancer START_LOAD_BALANCER=true
 	skipper make $(SKIPPER_PARAMS) _test_parallel
+	scripts/assisted_deployment.sh set_all_vips_dns
```
Comment (YuviGold) [inline Makefile:481]: why is that needed if `set_dns` is called as part of `preparing_for_installation`?

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 1 above)
Comment (ori-amizur) [reply discovery-infra/test_infra/helper_classes/cluster.py:634]: Terraform treats the file as resource.  It will be destroyed during teardown.

File: discovery-infra/test_infra/helper_classes/cluster.py [hunk 4]
```diff
@@ -621,6 +621,18 @@ def wait_for_install(self, timeout=consts.CLUSTER_INSTALLATION_TIMEOUT):
...
         )
 
+    def _set_dns(self):
+        cluster = self.api_client.cluster_get(self.id)
+        cluster_name = cluster.name
+        base_domain = cluster.base_dns_domain
+        main_cidr = self.nodes.controller.get_machine_cidr()
+        nameserver_ip = str(IPNetwork(main_cidr).ip + 1)
+        fname = f"/etc/NetworkManager/dnsmasq.d/openshift-{cluster_name}.conf"
+        contents = f"server=/api.{cluster_name}.{base_domain}/{nameserver_ip}\n" + f"server=/.apps.{cluster_name}.{base_domain}/{nameserver_ip}\n"
```
Comment (YuviGold) [inline discovery-infra/test_infra/helper_classes/cluster.py:631]: Using multi-line with `textwrap.dedent` would make it readable.

```suggestion
        from textwrap import dedent
        ....
        
        contents = dedent(f"""
        server=/api.{cluster_name}.{base_domain}/{nameserver_ip}
        server=/.apps.{cluster_name}.{base_domain}/{nameserver_ip}""")
```

File: Makefile (hunk 3 above)
Comment (ori-amizur) [reply Makefile:481]: To ask NM to reload

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 4 above)
Comment (ori-amizur) [reply discovery-infra/test_infra/helper_classes/cluster.py:631]: Is it more readable?

File: scripts/assisted_deployment.sh [hunk 5]
```diff
@@ -32,6 +32,16 @@ function set_dns() {
     echo "Finished setting dns"
 }
 
+function set_all_vips_dns() {
+    FILE="/etc/NetworkManager/conf.d/dnsmasq.conf"
+    if ! [ -f "${FILE}" ]; then
+        echo -e "[main]\ndns=dnsmasq" | sudo tee $FILE
+    fi
```
[the remark this answers is not in the document]
Comment (ori-amizur) [reply scripts/assisted_deployment.sh:39]: OK. Will refactor

File: skipper.yaml (hunk 2 above)
Comment (ori-amizur) [reply skipper.yaml:40]: OK

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 1 above)
Comment (ori-amizur) [reply discovery-infra/test_infra/helper_classes/cluster.py:634]: For cleanup after use.

File: Makefile (hunk 3 above)
Comment (YuviGold) [reply Makefile:481]: Could we reload inside `_set_dns()` function?

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 4 above)
Comment (YuviGold) [reply discovery-infra/test_infra/helper_classes/cluster.py:631]: Yes. You don't have one long line with strings concatenation nor new-line backslash.

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 1 above)
Comment (osherdp) [reply discovery-infra/test_infra/helper_classes/cluster.py:634]: and yet configuration will still hold until doing ``systemctl NetworkManager reload``, no?
I think it will become hard to investigate network problems

File: Makefile (hunk 3 above)
Comment (ori-amizur) [reply Makefile:481]: No.  It has to be done outside the container

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 1 above)
Comment (ori-amizur) [reply discovery-infra/test_infra/helper_classes/cluster.py:638]: It will be destroyed only when **_terraform destroy_** is called.
If you decide to leave the cluster alive, the configuration will remain in place.
During reload, if the cluster was destroyed, the operation will not have any effect (as expected).

Comment (ori-amizur) [reply discovery-infra/test_infra/helper_classes/cluster.py:638]: I think that leaving these files in this directory (part of the global NM) is incorrect behavior.

Comment (osherdp) [reply discovery-infra/test_infra/helper_classes/cluster.py:638]: yes, we might become inconsistent either way
anyway, I think it should be ok as is. It should only affect local environments and let's see what will actually happen on those environments

File: Makefile (hunk 3 above)
Comment (YuviGold) [reply Makefile:481]: Why? Can we make NetworkManager changes be applied from within as we do to the dnsmasq conf? It requires some socket maybe?

File: discovery-infra/test_infra/helper_classes/cluster.py (hunk 1 above)
Comment (YuviGold) [reply discovery-infra/test_infra/helper_classes/cluster.py:638]: I agree with @osherdp it might become a bit complicated to investigate... But we can try. sure, the TF cleanup is a great feature.

File: Makefile (hunk 3 above)
Comment (ori-amizur) [reply Makefile:481]: It is systemd interface.  Not sure we want to deal with it.  I don't think it is in the scope of this PR

**Review by YuviGold — APPROVED** [review]
LGTM
